### PR TITLE
Add test-coverage & codecov target and update circleci

### PR DESCRIPTION
**Pull request by vdemeester**
This is more or less what we used on some other docker project.

- Not sure if `test-coverage` should be on his own or be just `test`
- @dnephin I added `parellel` here too 👼 
- @nandhini915 the codecov token is set on circleci I guess ? 👼 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by dnephin** [comment]
Looks like it's failing because of missing dep.

Do you think we need this in the Makefile, or only the circleci?

I think parallel might actually slow it down, because we only get 1 CPU on CI.

I'm going to push some changes and you can let me know what you think.

**Comment by nandhini915** [comment]
@vdemeester I don't think the token would be required (since docker/cli is a public repo), but yes for private repos we'd setup codecov token in circleci.

**Comment by aaronlehmann** [comment]
We run codecov on PRs against the swarmkit and distribution repositories. I don't have anything against it. But I also can't recall a time when the coverage information has been useful. We've pushed back against PRs that didn't have adequate tests, but that came from reviewing the PR and noticing that the tests were insufficient. I don't think we've ever made the decision based on the `codecov` report. In the past we tried to automatically fail CI for PRs that decreased the coverage ratio, and this worked very poorly. There was random fluctuation in the results because of nondeterminism in some code and tests (for example, the order in which two goroutines execute might affect which code paths are taken). It also penalized changes that added extensive error checking, when those errors were not possible to trigger in practice (and therefore the code paths did not show up in code coverage).

cc @stevvooe

**Comment by dnephin** [comment]
I don't think we have any plans to make it fail CI, I agree that is unreliable.

For me the benefit to this integration is that it makes it easier to see the test coverage for your change. Test coverage may be easy to see in a PR that adds new features, but test coverage in a PR that refactor lots of code is less obvious.

The valuable metric to me is patch coverage, not total coverage.

File: scripts/test/unit [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+set -eu -o pipefail
+
+go test -tags daemon -v $@
```
Comment (stevvooe) [inline scripts/test/unit]: Why aren't these just in the Makefile? These kinds of systems are extremely hard to maintain.

File: scripts/test/unit-with-coverage [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+set -eu -o pipefail
+
+for pkg in $(go list ./... | grep -v /vendor/); do
+    ./scripts/test/unit \
+        -cover \
+        -coverprofile=profile.out \
+        -covermode=atomic \
+        ${pkg}
+    
+    if test -f profile.out; then
```
Comment (stevvooe) [inline scripts/test/unit-with-coverage]: Please include a comment as to why this required.

File: scripts/test/unit (hunk 1 above)
Comment (dnephin) [inline scripts/test/unit]: What are "these kinds of systems" ? And what makes them hard to maintain?

Comment (stevvooe) [inline scripts/test/unit]: A mess of bash scripts. It contributes to a horrid build experience and is very hard to modify.

This particular example is absurd. Why is this in its own file?

File: scripts/test/unit-with-coverage (hunk 2 above)
Comment (stevvooe) [inline scripts/test/unit-with-coverage]: Please address this. Why is this not an argument to the original command?

**Comment by stevvooe** [comment]
@vdemeester I would hope that you would at least address my comments before merging.

This Makefile setup gets a hard REJECT.

File: scripts/test/unit-with-coverage (hunk 2 above)
Comment (dnephin) [inline scripts/test/unit-with-coverage]: What is "this"? The `profile.out` filename? 

https://lk4d4.darth.io/posts/multicover/

**Comment by thaJeztah** [comment]
@dnephin @vdemeester can this PR be discussed so that things are addressed or consensus reached?

**Comment by dnephin** [comment]
Yes, I'm still trying to understand the concerns with this PR.   It was important to get it merged so that we have coverage on other PRs. If there are things that need to be changed, we can change them.

File: scripts/test/unit-with-coverage (hunk 2 above)
Comment (stevvooe) [inline scripts/test/unit-with-coverage]: The file is given as one argument, then copied to the other. Why not give the original file to the command and have it write directly? Are you trying to append it to a global coverage file?

Comment (dnephin) [inline scripts/test/unit-with-coverage]: The file isn't copied. The file is appended to the global coverage file yes. That's what it's doing.

**Comment by thaJeztah** [comment]
This didn't make it into 17.06 - moving milestone

**Pull request merged**